// pr0gramm-api/lib/common-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemFlags = exports.Vote = exports.UserMark = void 0;
var UserMark;
(function (UserMark) {
    UserMark[UserMark["Schwuchtel"] = 0] = "Schwuchtel";
    UserMark[UserMark["Neuschwuchtel"] = 1] = "Neuschwuchtel";
    UserMark[UserMark["Altschwuchtel"] = 2] = "Altschwuchtel";
    UserMark[UserMark["Administrator"] = 3] = "Administrator";
    UserMark[UserMark["Gebannt"] = 4] = "Gebannt";
    UserMark[UserMark["Moderator"] = 5] = "Moderator";
    UserMark[UserMark["Fliesentisch"] = 6] = "Fliesentisch";
    UserMark[UserMark["LebendeLegende"] = 7] = "LebendeLegende";
    UserMark[UserMark["Wichtel"] = 8] = "Wichtel";
    UserMark[UserMark["EdlerSpender"] = 9] = "EdlerSpender";
    UserMark[UserMark["Mittelaltschwuchtel"] = 10] = "Mittelaltschwuchtel";
    UserMark[UserMark["Altmoderator"] = 11] = "Altmoderator";
    UserMark[UserMark["Communityhelfer"] = 12] = "Communityhelfer";
    UserMark[UserMark["Nutzerbot"] = 13] = "Nutzerbot";
    UserMark[UserMark["Systembot"] = 14] = "Systembot";
})(UserMark = exports.UserMark || (exports.UserMark = {}));
var Vote;
(function (Vote) {
    Vote[Vote["Down"] = -1] = "Down";
    Vote[Vote["None"] = 0] = "None";
    Vote[Vote["Up"] = 1] = "Up";
    Vote[Vote["Favorite"] = 2] = "Favorite";
})(Vote = exports.Vote || (exports.Vote = {}));
var ItemFlags;
(function (ItemFlags) {
    ItemFlags[ItemFlags["SFW"] = 1] = "SFW";
    ItemFlags[ItemFlags["NSFW"] = 2] = "NSFW";
    ItemFlags[ItemFlags["NSFL"] = 4] = "NSFL";
    ItemFlags[ItemFlags["NSFP"] = 8] = "NSFP";
    ItemFlags[ItemFlags["All"] = 15] = "All";
})(ItemFlags = exports.ItemFlags || (exports.ItemFlags = {}));
