// pr0gramm-api/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addApiKeyToHeader = exports.encodeQueryParams = exports.addQueryParams = exports.createDefaultHeaders = exports.createTagList = exports.ensureUnixTimestamp = void 0;
const client_constants_1 = require("./client-constants");
function ensureUnixTimestamp(v) {
    "use asm";
    return typeof v === "number"
        ? v | 0
        : (v.getTime() / 1000) | 0;
}
exports.ensureUnixTimestamp = ensureUnixTimestamp;
function createTagList(tags) {
    return tags.map(tag => tag.split(",")).flat().join(",");
}
exports.createTagList = createTagList;
function createDefaultHeaders() {
    return {
        "User-Agent": (0, client_constants_1.getUserAgent)(),
    };
}
exports.createDefaultHeaders = createDefaultHeaders;
function addQueryParams(url, params) {
    return params
        ? url + "?" + encodeQueryParams(params)
        : url;
}
exports.addQueryParams = addQueryParams;
function encodeQueryParams(params) {
    return params === undefined
        ? ""
        : new URLSearchParams(removeUndefinedValues(params)).toString();
}
exports.encodeQueryParams = encodeQueryParams;
function removeUndefinedValues(qp) {
    const copy = Object.assign({}, qp);
    Object.keys(copy).forEach((key) => (copy[key] === undefined) && delete copy[key]);
    return copy;
}
function addApiKeyToHeader(headers, oAuthKey) {
    if (!oAuthKey)
        return headers;
    return Object.assign(Object.assign({}, headers), { Authorization: `Bearer ${oAuthKey}` });
}
exports.addApiKeyToHeader = addApiKeyToHeader;
